File: robotmasters/tests/spawn.ts
import { Program } from '@coral-xyz/anchor'
import * as anchor from '@coral-xyz/anchor'
import { Robotmasters } from '../target/types/robotmasters'

import { expect } from 'chai'
import { deriveComponentPda } from './utils'
import { spawnControlPda, spawnPda, itemAuthority } from './0_init'

describe('Spawn', () => {
  const provider = anchor.AnchorProvider.env()
  anchor.setProvider(provider)

  const program = anchor.workspace.Robotmasters as Program<Robotmasters>

  it('Create a Spawn', async () => {
    await program.methods
      .createSpawn({
        healthCap: 1,
        duration: 120,
        critChance: 100,
        critMultiplier: 100,
        damageBase: 5,
        damageRange: 10,
        destroyOnCollision: true,
        element: 1,
        height: 3,
        width: 6,
        outputX: 14,
        outputY: 8,
        args: [0, 0, 0, 0],
        script: Buffer.from([]),
      })
      .accounts({
        authority: program.provider.publicKey,
        spawn: spawnPda,
        control: spawnControlPda,
      })
      .rpc()

    const spawnData = await program.account.spawn.fetch(spawnPda)

    expect(typeof spawnData.element.punct).eq('object')

    const spawnControlData = await program.account.componentControl.fetch(
      spawnControlPda
    )

    expect(spawnControlData.counter).eq(0)
    expect(program.provider.publicKey.equals(spawnControlData.owner)).eq(true)
  })

  it('Update a draft state Spawn', async () => {
    await program.methods
      .updateSpawn({
        healthCap: 1,
        duration: 120,
        critChance: 100,
        critMultiplier: 100,
        damageBase: 5,
        damageRange: 10,
        destroyOnCollision: true,
        element: 2,
        height: 3,
        width: 6,
        outputX: 14,
        outputY: 8,
        args: [0, 0, 0, 0],
        script: Buffer.from([]),
        state: { draft: {} },
      })
      .accounts({
        spawn: spawnPda,
        control: spawnControlPda,
        signer: program.provider.publicKey,
      })
      .rpc()

    const spawnData = await program.account.spawn.fetch(spawnPda)

    expect(typeof spawnData.element.blast).eq('object')
  })

  it('Update the spawn to pending state', async () => {
    await program.methods
      .updateSpawn({
        healthCap: 1,
        duration: 120,
        critChance: 100,
        critMultiplier: 100,
        damageBase: 5,
        damageRange: 10,
        destroyOnCollision: true,
        element: 2,
        height: 3,
        width: 6,
        outputX: 14,
        outputY: 8,
        args: [0, 0, 0, 0],
        script: Buffer.from([]),
        state: { pending: {} },
      })
      .accounts({
        spawn: spawnPda,
        control: spawnControlPda,
        signer: program.provider.publicKey,
      })
      .rpc()
    const spawnData = await program.account.spawn.fetch(spawnPda)
    expect(typeof spawnData.state.pending).eq('object')
  })

  it('Should not update a Spawn when not in Draft state', async () => {
    try {
      await program.methods
        .updateSpawn({
          healthCap: 10,
          duration: 120,
          critChance: 100,
          critMultiplier: 100,
          damageBase: 5,
          damageRange: 10,
          destroyOnCollision: true,
          element: 2,
          height: 3,
          width: 6,
          outputX: 14,
          outputY: 8,
          args: [0, 0, 0, 0],
          script: Buffer.from([]),
          state: { pending: {} },
        })
        .accounts({
          spawn: spawnPda,
          control: spawnControlPda,
          signer: program.provider.publicKey,
        })
        .rpc()
      expect.fail('--')
    } catch (e) {
      expect(e.message).to.not.eq('--')
    }
  })

  it('Admin should be able to transition Spawn to other state', async () => {
    await program.methods
      .updateSpawn({
        healthCap: 1,
        duration: 120,
        critChance: 100,
        critMultiplier: 100,
        damageBase: 5,
        damageRange: 10,
        destroyOnCollision: true,
        element: 2,
        height: 3,
        width: 6,
        outputX: 14,
        outputY: 8,
        args: [0, 0, 0, 0],
        script: Buffer.from([]),
        state: { published: {} },
      })
      .accounts({
        spawn: spawnPda,
        control: spawnControlPda,
        signer: itemAuthority.publicKey,
      })
      .signers([itemAuthority])
      .rpc()
    const spawnData = await program.account.spawn.fetch(spawnPda)
    expect(typeof spawnData.state.published).eq('object')
  })

  it('Should be able to make a version of the Spawn', async () => {
    const [newSpawnPda] = deriveComponentPda(program.programId, 'spawn', 0, 1)

    await program.methods
      .versionSpawn({
        healthCap: 1,
        duration: 120,
        critChance: 100,
        critMultiplier: 100,
        damageBase: 5,
        damageRange: 10,
        destroyOnCollision: true,
        element: 1,
        height: 3,
        width: 6,
        outputX: 14,
        outputY: 8,
        args: [0, 0, 0, 0],
        script: Buffer.from([]),
      })
      .accounts({
        oldSpawn: spawnPda,
        newSpawn: newSpawnPda,
        control: spawnControlPda,
        signer: program.provider.wallet.publicKey,
      })
      .rpc()

    const spawnControlData = await program.account.componentControl.fetch(
      spawnControlPda
    )

    expect(spawnControlData.counter).eq(1)

    const spawnData = await program.account.spawn.fetch(newSpawnPda)
    expect(spawnData.version).eq(spawnControlData.counter)
  })
})

File: robotmasters/programs/robotmasters/src/states/spawn.rs
use bolt_lang::prelude::*;

use super::{ComponentState, Element, FromAccountInfo};

#[account]
pub struct Spawn {
    pub bump: u8,
    pub id: u32,
    pub version: u32,
    pub state: ComponentState,

    pub health_cap: u8,
    pub duration: u16,
    pub damage_base: u8,
    pub damage_range: u8,
    pub crit_chance: u8,
    pub crit_multiplier: u8,
    pub element: Element,
    pub destroy_on_collision: bool,
    pub width: u8,
    pub height: u8,
    pub output_x: u8,
    pub output_y: u8,
    pub args: [Option<u8>; 4],
    pub script: Vec<u8>,
    pub spawns: Vec<u32>,
}

impl Spawn {
    pub fn len(script_len: usize, spawns_len: usize) -> usize {
        8 +  // account discriminator
		1 +  // bump
		1 +  // state
		4 +  // id
		4 +  // version
        1 +  // health_cap
        2 +  // duration
        1 +  // damage_base
        1 +  // damage_range
        1 +  // crit_chance
        1 +  // crit_multiplier
        1 +  // element
        1 +  // destroy_on_collision
        1 +  // width
        1 +  // height
        1 +  // output_x
        1 +  // output_y
		4 * (1 + 1) +  // args
		(4 + script_len) +
		(4 + (4 * spawns_len))
    }

    pub fn serialize(&self, args: [Option<u8>; 4]) -> Vec<u16> {
        let output_len = 12 + 4 + self.script.len();
        let mut output = Vec::with_capacity(output_len);

        output.push(self.health_cap as u16);
        output.push(self.duration);
        output.push(self.damage_base as u16);
        output.push(self.damage_range as u16);
        output.push(self.crit_chance as u16);
        output.push(self.crit_multiplier as u16);
        output.push(self.element.to_byte() as u16);
        output.push((if self.destroy_on_collision { 1 } else { 0 }) as u16);
        output.push(self.width as u16);
        output.push(self.height as u16);
        output.push(self.output_x as u16);
        output.push(self.output_y as u16);
        output.extend(args.iter().map(|opt| opt.unwrap_or(0) as u16));
        output.extend(self.script.iter().map(|&x| x as u16));

        output
    }
}

impl FromAccountInfo for Spawn {}
